src/dmd/dcast.d

```diff
@@ -3056,8 +3056,6 @@ Lagain:
     {
         if (t1.mod != t2.mod)
         {
-            if (!t1.isImmutable() && !t2.isImmutable() && t1.isShared() != t2.isShared())
-                goto Lincompatible;
```

Comment:
why is immutable-mutable ok, immutable-shared ok, but shared-mutable not ok?

Reply:
commenting the if doesn't seem to influence any code in the test suite.

Reply:
Did you try to implement a custom, non synchronized `opEquals` ?